torch/csrc/jit/tensorexpr/block_codegen.cpp

```diff
@@ -0,0 +1,319 @@
+    os() << (block_idx ? "block: bs_N}" : "block: bs_DPE}");
+    ++trip;
+  }
+  os() << ")";
+}
+
+void BlockPrinter::PrintReshapeInfo(
+    std::unordered_set<const Buf*> bufs,
+    bool reverse) {
+  for (const auto& buf : bufs) {
+    emitIndent();
+    os() << "reshape("
+         << (reverse ? block_analysis_->getFlatInputName(buf)
+                     : block_analysis_->getInputName(buf))
+         << ", "
+         << (reverse ? block_analysis_->getInputName(buf)
+                     : block_analysis_->getFlatInputName(buf))
+         << ")" << std::endl;
+  }
+}
+
+void BlockPrinter::PrintDMAs(std::unordered_set<const Buf*> bufs) {
+  for (const auto& read : bufs) {
+    emitIndent();
+    os() << "dma_in(";
+    os() << block_analysis_->getFlatInputName(read);
+    os() << ")" << std::endl;
+  }
+}
+void BlockPrinter::PrintAdjustBuffers(std::unordered_set<const Buf*> bufs) {
+  for (const auto& read : bufs) {
+    emitIndent();
+    os() << "adjust_buffer(";
+    os() << block_analysis_->getFlatInputName(read);
+    os() << ")" << std::endl;
+  }
+}
+
+void BlockPrinter::visit(const Load* v) {
+  os() << block_analysis_->getFlatInputName(v->buf()) << ".buffer, ";
+}
+void BlockPrinter::visit(const Store* v) {
+  emitIndent();
+  os() << *v->value() << block_analysis_->getFlatInputName(v->buf())
+       << ".tensor)" << std::endl;
+}
+
+void BlockPrinter::visit(const Block* v) {
+  os() << "{" << std::endl;
+  indent_++;
+  for (Stmt* s : v->stmts()) {
+    s->accept(this);
+  }
+  indent_--;
+  emitIndent();
+  os() << "}";
+}
+
+std::string BlockCodeGen::GetUniqueFuncName(const std::string& func_prefix) {
+  // We are using a global counter here to make sure difference instances
+  // within BlockCodeGen have different names.
+  static int64_t counter = 0;
+  ++counter;
+  int64_t value = counter;
+  return func_prefix + "_" + std::to_string(value);
+}
+
+void BlockCodeGen::Initialize(
+    std::unique_ptr<BlockAnalysis> kernel_block_analysis) {
+  block_analysis_ = std::move(kernel_block_analysis);
+  printer_ = std::make_unique<BlockPrinter>(&oss_, block_analysis_.get());
+
+  Stmt* stmt_v = stmt();
+  stmt_v->accept(block_analysis_.get());
+
+  auto buf_reads = block_analysis_->reads();
+  auto buf_writes = block_analysis_->stores();
+  // Ensure all Bufs in reads/writes are in the map
+  std::unordered_set<const Buf*> bufs(buf_reads.begin(), buf_reads.end());
+  bufs.insert(buf_writes.begin(), buf_writes.end());
+  if (!block_analysis_->areBufsInMap(bufs)) {
+    throw std::runtime_error("BlockCodeGen: Entry not in input/Buffer map");
+  };
+
+  std::string func_name = GetUniqueFuncName("func");
+  os() << "kernel " << func_name << "(";
+  for (auto const& arg : buf_writes) {
+    os() << block_analysis_->getInputName(arg);
+  }
+  for (auto const& arg : buf_reads) {
+    os() << ";" << block_analysis_->getInputName(arg);
+  }
+  os() << ")";
+
+  stmt_v->accept(printer_.get());
+
+#if DEBUG_PRINT
```

Comment:
Please use `GRAPH_DEBUG`/`GRAPH_DUMP`/`GRAPH_UPDATE` instead of adhoc debug prints. I know DEBUG_PRINT is used in cuda codegen, but it's because we've been lazy to switch it to the proper logging, not because it should be that way :)

Reply:
done!